rolling-shutter/medley/broker.go

```diff
@@ -0,0 +1,189 @@
+package medley
+
+import (
+	"sync"
+
+	"github.com/rs/zerolog/log"
+)
+
+// Broker allows to distribute a value to multiple
+// receiving channels, so that all channels receive the same value.
+// This is different to the Fan-Out pattern where multiple channels
+// are listening for a send on one channel, but only the first channel
+// to select the receive operation receives this value.
+//
+// The Broker allows to set whether the internal send operation is
+// blocking or non blocking. In the blocking case, all receiving
+// (subscribed) channels have to successively select the receive in order
+// for the broker to continue operation. So if one receiver blocks,
+// everything else will block as well.
+// In the non-blocking case, a subscribed channel that is blocking
+// while the published value is sent will miss the value.
+//
+// One has to be careful in when to select which case.
+// The blocking behavior can be partially subverted by increasing
+// the subscribers buffer size - but if a subscriber is blocked forever,
+// this will only work until this subscribers buffer is full.
+type Broker[T any] struct {
+	mux                  sync.RWMutex
+	stop                 chan struct{}
+	publish              chan T
+	subscribe            chan chan T
+	unsubscribe          chan chan T
+	initialSubscriptions []chan T
+	running              bool
+	nonBlockingSend      bool
+}
+
+func NewBroker[T any](nonBlocking bool) *Broker[T] {
+	return &Broker[T]{
+		stop:                 make(chan struct{}),
+		publish:              make(chan T, 1),
+		subscribe:            make(chan chan T),
+		unsubscribe:          make(chan chan T),
+		nonBlockingSend:      nonBlocking,
+		initialSubscriptions: []chan T{},
+		running:              false,
+	}
+}
+
```

Comment:
The broker code can be much simplified if you replace `NewBroker` and the `Start` method with a function `StartNewBroker` that creates a broker and starts the go routine. You can get rid of the `running` flag, the `mux` field as well as the `initialSubscriptions` field (if there is a need, this can be a argument to `StartNewBroker` and the related logic. There's no need for locking since the struct becomes immutable.